rpcs3/Emu/Cell/Modules/cellMic.cpp

```diff
@@ -134,48 +134,109 @@ void mic_context::operator()()
 	// Cleanup
 	for (auto& mic_entry : mic_list)
 	{
-		mic_entry.second.close_microphone();
+		mic_entry.close_microphone();
 	}
 }
 
 void mic_context::load_config_and_init()
 {
-	auto device_list = fmt::split(g_cfg.audio.microphone_devices.to_string(), {"@@@"});
+	mic_list = {};
 
-	if (!device_list.empty() && mic_list.empty())
+	const std::vector<std::string> device_list = fmt::split(g_cfg.audio.microphone_devices.to_string(), {"@@@"});
+
+	if (!device_list.empty())
 	{
-		switch (g_cfg.audio.microphone_type)
-		{
-		case microphone_handler::standard:
+		// We only register the first device. The rest is registered with cellAudioInRegisterDevice.
+		if (g_cfg.audio.microphone_type == microphone_handler::singstar)
 		{
-			for (s32 index = 0; index < static_cast<s32>(device_list.size()); index++)
+			microphone_device& device = ::at32(mic_list, 0);
+			device = microphone_device(microphone_handler::singstar);
+			device.set_registered(true);
+			device.add_device(device_list[0]);
+
+			// Singstar uses the same device for 2 players
+			if (device_list.size() >= 2)
 			{
-				mic_list.emplace(std::piecewise_construct, std::forward_as_tuple(index), std::forward_as_tuple(microphone_handler::standard));
-				::at32(mic_list, index).add_device(device_list[index]);
+				device.add_device(device_list[1]);
 			}
-			break;
 		}
-		case microphone_handler::singstar:
+		else
 		{
-			mic_list.emplace(std::piecewise_construct, std::forward_as_tuple(0), std::forward_as_tuple(microphone_handler::singstar));
-			::at32(mic_list, 0).add_device(device_list[0]);
-			if (device_list.size() >= 2)
-				::at32(mic_list, 0).add_device(device_list[1]);
-			break;
+			[[maybe_unused]] const u32 index = register_device(device_list[0]);
 		}
-		case microphone_handler::real_singstar:
-		case microphone_handler::rocksmith:
+	}
+}
+
+u32 mic_context::register_device(const std::string& device_name, const std::string& name_2_for_singstar)
+{
+	usz index = 0;
+	for (usz i = 0; i < mic_list.size(); i++)
+	{
+		microphone_device& device = ::at32(mic_list, i);
+		if (!device.is_registered())
 		{
-			mic_list.emplace(std::piecewise_construct, std::forward_as_tuple(0), std::forward_as_tuple(g_cfg.audio.microphone_type));
-			::at32(mic_list, 0).add_device(device_list[0]);
+			index = i;
 			break;
 		}
-		case microphone_handler::null:
-		default: break;
+	}
+
+	// TODO: Check max mics properly
+	ensure(index <= mic_list.size(), "cellMic max mics exceeded during registration");
```

Comment:
It should be < mic_list.size() also with current logic if no free spot is found index will be 0 which is not correct.